### PR TITLE
[SMTChecker] Implement support for memory allocation

### LINKED ISSUE #6050: [SMTChecker] Support memory array allocation
For example:
```
Warning: Assertion checker does not yet implement this type of function call.
  --> Blake2bTest.sol:41:32:
   |
41 |         bytes memory scratch = new bytes(128);
   |                                ^^^^^^^^^^^^^^
```

### PR DESCRIPTION
**Pull request by axic**
Fixes #6050.

### CODE REVIEW

File: libsolidity/formal/SMTEncoder.cpp [hunk 1]
```diff
@@ -663,6 +663,17 @@ void SMTEncoder::endVisit(FunctionCall const& _funCall)
 	case FunctionType::Kind::Event:
 		// These can be safely ignored.
 		break;
+	case FunctionType::Kind::ObjectCreation:
+	{
+		ArrayType const& arrayType = dynamic_cast<ArrayType const&>(*_funCall.annotation().type);
+		// TODO: support other kinds of arrays
```
Comment (axic) [inline libsolidity/formal/SMTEncoder.cpp]: I actually think this works for any kind of array, as it just presets them to zero and sets the length.

File: test/libsolidity/smtCheckerTests/operators/bytes_new.sol [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+		assert(x[2] == 0);
+	}
+	function h() public pure {
+		bytes memory x = new bytes(3);
+		assert(x.length == 3);
+		x[0] = 0x12;
+		x[1] = 0x34;
+		assert(x[0] == 0x12);
+		assert(x[1] == 0x34);
+		// This should be an out-of-bounds assertion.
```
Comment (axic) [inline test/libsolidity/smtCheckerTests/operators/bytes_new.sol:22]: @leonardoalt the fixed bytes indexing should take the length into account.

File: libsolidity/formal/SMTEncoder.cpp (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp]: Yes it works for any Solidity array. Maybe you can already change that here?

File: libsolidity/formal/SMTEncoder.cpp [hunk 3]
```diff
@@ -663,6 +663,17 @@ void SMTEncoder::endVisit(FunctionCall const& _funCall)
...
+		if (arrayType.isByteArray())
+		{
+			visitObjectCreation(_funCall);
+			return;
+		}
+		[[fallthrough]];
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]: Why fallthrough?

File: libsolidity/formal/SMTEncoder.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/formal/SMTEncoder.cpp]: So just remove the warning and consider it working for everything?

File: libsolidity/formal/SMTEncoder.cpp (hunk 3 above)
Comment (axic) [reply libsolidity/formal/SMTEncoder.cpp]: To have the warning displayed?

File: libsolidity/formal/SMTEncoder.cpp [hunk 4]
```diff
@@ -735,6 +746,26 @@ void SMTEncoder::visitGasLeft(FunctionCall const& _funCall)
...
 }
 
+void SMTEncoder::visitObjectCreation(FunctionCall const& _funCall)
+{
+	auto const& args = _funCall.arguments();
+	solAssert(args.size() >= 1, "");
+	auto argType = args.front()->annotation().type->category();
+	solAssert(argType == Type::Category::Integer || argType == Type::Category::RationalNumber, "");
+
+	// TODO: support the special case of integer literals to have less symbolic variables, but this works too
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]: I think it's fine like this, don't think it's gonna help much to have it separate.

Comment (axic) [reply libsolidity/formal/SMTEncoder.cpp]: It is actually 3 more lines of code, but we should consider that if solving is becoming too slow and we have enough heavy examples.

**Review by leonardoalt — COMMENTED** [review]
Really nice!
I think it can already be extended to all array types though

File: libsolidity/formal/SMTEncoder.cpp (hunk 4 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp]: But why would it be easier in that case?

File: libsolidity/formal/SMTEncoder.cpp (hunk 3 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp]: Ah, for the else case. Was already thinking about the version that supports all arrays :p

File: libsolidity/formal/SMTEncoder.cpp (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp]: Is `ObjectCreation` only for arrays? I think so, right? struct constructors go into function call

Comment (axic) [reply libsolidity/formal/SMTEncoder.cpp]: Struct constructors are separate.

File: libsolidity/formal/SMTEncoder.cpp (hunk 4 above)
Comment (axic) [reply libsolidity/formal/SMTEncoder.cpp]: It is two useless assertions less (`setSymbolicUnknownValue` adds a min/max).

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp]: ah right

**Comment by axic** [comment]
@leonardoalt changed it, I'll squash before merge.

**Review by leonardoalt — COMMENTED** [review]
Maybe a few more tests with other types of arrays?
Otherwise looks good.

File: test/libsolidity/smtCheckerTests/types/function_type_arrays.sol [hunk 5]
```diff
@@ -1,4 +1,4 @@
-pragma experimental SMTChecker;
+rpragma experimental SMTChecker;
```
Comment (leonardoalt) [inline test/libsolidity/smtCheckerTests/types/function_type_arrays.sol]: typo

**Review by leonardoalt — DISMISSED** [review]

**Review by leonardoalt — APPROVED** [review]

File: libsolidity/formal/SMTEncoder.cpp (hunk 4 above)
Comment (axic) [reply libsolidity/formal/SMTEncoder.cpp]: We could also add an optimizer step removing such assertions :wink:

**Pull request merged**